src/preamble/02_what_is_a_module_and_a_package.md

```diff
@@ -1 +1,365 @@
+[chapter](../getting_fancy/06_foreign_libraries.md) on them.
+
+It is not necessary to commit these fields to memory, but it is good to be aware
+of them. Hopefully when you look at the layout of a `.cabal` file next, you will
+see the `package <- component <- module` hierarchy reflected in the nested
+fields!
+
+## What is a package?
+
+If a module is the smallest unit of code in cabal, a **package** is the largest
+unit of distribution. At its core, all a package is is a distributable artefact
+that provides access to components.
+
+> You might remember cabal projects from earlier, as being _upstream_ of
+> packages. Good catch, a **project** is bigger in the sense that it is a
+> grouping of packages, but it is not distinct in terms of the distribution of
+> code.
+
+The `.cabal` file represents a single package, and is really just a collection
+of metadata about the packages constituent parts. It defines the packages
+components and their internal dependencies (list of modules) as well as the
+packages external dependencies; usually other packages.
+
+To make things more concrete, here is a pseudo-Haskell type representing the
+package-component-module hierarchy:
+
+```haskell
+data Module = Module { moduleName :: Text, moduleEntities :: [Entities] }
+
+data ComponentType = Executable | Library | TestSuite
+
+data Component =
+  Component
+    { componentName    :: Text
+    , componentType    :: ComponentType
+    , componentModules :: [Module]
+    }
+
+type Package = [Component]
+```
+
+> Note: technically a package can contain no components, but if you try and run
+> `cabal build` you will get an error telling you that your package contains no
+> components.
+
+## Summary
+
+Cabal is a tool for defining and building Haskell packages. Even if we don't
+intend to distribute our code, we can still think of it as an unrealized
+package. This is important because, inherent in the definition of a package, are
+all of its dependencies. This is a very common use case for cabal; to bring a
+bunch of dependencies in to scope so that we can explore them and call them from
+our own code (even if we don't intend to do anything with our code). This
+exploratory workflow is facilitated by commands like `cabal repl`, and
+`cabal env` (which is being worked on currently).
+
+Therefore it is informative to think of cabal as handling several distinct
+concerns:
+
+1. Describing the internal dependency structure of a package in terms of
+   modules.
+2. Managing the external dependencies of our project and allowing us to access
+   them from within our modules.
+3. Managing the building and distribution of our package, which is really just a
+   collection of components.
+
+I believe that the reader of this guide will get the most out of it if they
+consider which of these functions they are interested in performing with Cabal.
+Hopefully understanding these distinct considerations will reveal some of the
+intention behind subsequent sections!
```

Comment:
Great stuff. Generally, a very generous chapter. A complete opposite of dev docs. Well done.